File: src/main/java/com/extendedae_plus/network/CPacketUploadTerminalPattern.java
package com.extendedae_plus.network;

import appeng.menu.me.items.PatternEncodingTermMenu;
import com.extendedae_plus.common.impl.pattern.PatternUploader;
import com.extendedae_plus.network.base.CPacketGeneric;
import com.extendedae_plus.network.base.EAEPNetworkPacket;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.server.level.ServerPlayer;

@EAEPNetworkPacket("upload_terminal_pattern")
public record CPacketUploadTerminalPattern(int serverID) implements CPacketGeneric {
    public static final StreamCodec<FriendlyByteBuf, CPacketUploadTerminalPattern> STREAM_CODEC = StreamCodec.composite(
            ByteBufCodecs.INT, CPacketUploadTerminalPattern::serverID,
            CPacketUploadTerminalPattern::new
    );

    @Override
    public void handleServer(ServerPlayer player) {
        if (!(player.containerMenu instanceof PatternEncodingTermMenu menu)) return;

        PatternUploader.uploadToProvider(menu, this.serverID);
    }
}

File: src/main/kotlin/com/extendedae_plus/network/CPacketUploadTerminalPattern.kt
package com.extendedae_plus.network

import appeng.menu.me.items.PatternEncodingTermMenu
import com.extendedae_plus.common.impl.pattern.PatternUploader
import com.fish.fishlib.network.FishNetworkPacket
import com.fish.fishlib.network.PacketStreamCodec
import com.fish.fishlib.network.base.CPacketGeneric
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.server.level.ServerPlayer

@FishNetworkPacket("upload_terminal_pattern")
@JvmRecord
data class CPacketUploadTerminalPattern(val serverID: Int) : CPacketGeneric {
    override fun handleServer(player: ServerPlayer) {
        PatternUploader.uploadFromMenuEncoding(
            player.containerMenu as? PatternEncodingTermMenu ?: return,
            this.serverID
        )
    }

    companion object {
        @PacketStreamCodec
        val streamCodec: StreamCodec<FriendlyByteBuf, CPacketUploadTerminalPattern> =
            StreamCodec.composite(
                ByteBufCodecs.INT, CPacketUploadTerminalPattern::serverID,
                ::CPacketUploadTerminalPattern
            )
    }
}
